core/consensus/src/trace.rs

```diff
@@ -0,0 +1,43 @@
+use log::{error, trace};
+use moodyblues_sdk::time::now;
+use moodyblues_sdk::trace::{set_boxed_tracer, Metadata, Trace, TracePoint};
+use serde_json::to_string;
+
+use protocol::{ProtocolError, ProtocolResult};
+
+use crate::ConsensusError;
+
+struct MetricTracer {
+    address: String,
+}
+
+impl MetricTracer {
+    fn new(address: String) -> MetricTracer {
+        MetricTracer { address }
+    }
+}
+
+impl Trace for MetricTracer {
+    fn report(&self, point: TracePoint) {
+        match to_string(&point) {
+            Ok(json) => trace!(target: "metrics", "{}", json),
+            Err(e) => error!("tracing: convert json error {:?}", e),
+        }
+    }
+
+    fn metadata(&self) -> Metadata {
+        Metadata {
+            address: self.address.clone(),
+        }
+    }
+
+    fn now(&self) -> u64 {
+        now()
+    }
+}
+
+pub fn init_tracer(address: String) -> ProtocolResult<()> {
+    set_boxed_tracer(Box::new(MetricTracer::new(address))).map_err(|_| {
+        ProtocolError::from(ConsensusError::Other("failed to init tracer ".to_string()))
```

Comment:
ConsensusError::Other("failed to init tracer ".to_string().into

Reply:
map_err is convenient here, why still use `Err into`

Reply:
`ProtocolError::from` is unnecessary.

Try replacing it with `ConsensusError::Other("failed to init tracer ".to_string()).into`

Reply:
OK, I know, this looks better